piff/gp_interp_2pcf.py

```diff
@@ -0,0 +1,375 @@
+        MIN = np.sqrt(1./rho)
+        MAX = np.sqrt(size_x**2 + size_y**2)/2.
+    
+        if y_err is None or np.sum(y_err) == 0 :
+            cat = treecorr.Catalog(x=X[:,0], y=X[:,1], k=(y-np.mean(y)))
+        else:
+            cat = treecorr.Catalog(x=X[:,0], y=X[:,1], k=(y-np.mean(y)), w=1./y_err**2)
+        kk = treecorr.KKCorrelation(min_sep=MIN, max_sep=MAX, nbins=20)
+        kk.process(cat)
+
+        distance = np.exp(kk.logr)
+        Coord = np.array([distance,np.zeros_like(distance)]).T
+
+        def PCF(param,k=kernel):
+            kernel =  k.clone_with_theta(param)
+            pcf = kernel.__call__(Coord,Y=np.zeros_like(Coord))[:,0]
+            return pcf
+
+        def chi2(param,disp=np.std(y)):
+            residual = kk.xi - PCF(param)
+            var = disp**2
+            return np.sum(residual**2/var)
+
+        p0 = kernel.theta
+        results_fmin = op.fmin(chi2,p0,disp=False)
+        results_bfgs = op.minimize(chi2,p0,method="L-BFGS-B")
+        results = [results_fmin, results_bfgs['x']]
+        chi2_min = [chi2(results[0]), chi2(results[1])]
+        ind_min = chi2_min.index(min(chi2_min))
+        results = results[ind_min]
+
+        self._2pcf.append(kk.xi)
+        self._2pcf_dist.append(distance)
+        kernel =  kernel.clone_with_theta(results)
+        self._2pcf_fit.append(PCF(kernel.theta))
+        return kernel
+
+    def _predict(self, Xstar):
+        """ Predict responses given covariates.
+        :param Xstar:  The independent covariates at which to interpolate.  (n_samples, n_features).
+        :returns:  Regressed parameters  (n_samples, n_targets)
+        """
+        if self.npca>0:
+            y_init = self._y_pca
+            y_err = self._y_pca_err
+        else:
+            y_init = self._y
+            y_err = self._y_err
+
+        ystar = np.array([self.return_gp_predict(y_init[:,i]-self._mean[i], self._X, Xstar, ker, y_err=y_err[:,i]) for i,ker in enumerate(self.kernels)]).T
+        
+        for i in range(self.nparams):
+            ystar[:,i] += self._mean[i]
+        if self.npca > 0:
+            ystar = self._pca.inverse_transform(ystar)
+        return ystar
+
+    def return_gp_predict(self,y, X1, X2, kernel, y_err=None):
+        """Compute interpolation with gaussian process for a given kernel.
+
+        :param y:  The dependent responses.  (n_samples, n_targets)
+        :param X1:  The independent covariates.  (n_samples, n_features)
+        :param X2:  The independent covariates at which to interpolate.  (n_samples, n_features)
+        :param kernel: sklearn.gaussian_process kernel.
+        :param y_err: Error of y. (n_samples, n_targets)
+        """
+
+        if y_err is None:
+            y_err = np.zeros_like(y)
+        HT = kernel.__call__(X2,Y=X1)
+        K = kernel.__call__(X1) + np.eye(len(y))*y_err**2
+        factor = (cholesky(K, overwrite_a=True, lower=False), False)
+        alpha = cho_solve(factor, y, overwrite_b=False)
```

Comment:
Based on the way this is written, I think the majority of times this function is called it will be using the same X1, y, y_err from when the model is fitted. This means you wouldn't have to keep recomputing alpha, which probably would help speed things up.

Reply:
I guess this is true only if you are using the same kernel and if you have exactly same hyperparameters (so which means that you are not optimizing your hyperparameters). In this case this is the same alpha for all PSF parameters. However, in the general case you have one alpha per PSF parameter.